### DISCUSSION TITLE
AI assistant for adjusting Dandiset metadata

### DISCUSSION THREAD

**Discussion by magland** [flatironinstitute/neurosift#215] [General]
From suggestion of @satra, exploring the possibility of a chat interface for setting Dandiset metadata. Could take care of some of the more painstaking aspects such as extracting authors from a publication, target brain structures derived from uberon, or extracting information from the dandiset.

I made a test page to show how this could be done
https://neurosift.app/?p=/test&dandisetId=000722

with a sample Dandiset 000722,

You can talk to that assistant on the main page, but it can't yet edit things directly (but it *can* probe information about the Dandiset and knows its limitations).

But you can use a chat to edit the Contributors.

Click on the "Contributors" link in the right panel. Then ask it to do things like:

"Remove Chiara" (sorry Chiara if you ever read this :))

"Correct spelling errors" (there actually is a typo in this dataset)

"Add authors from this citation [paste]"

"Please reformat the names to conform to last, first"

![image](https://github.com/user-attachments/assets/6ecae0ef-9282-4590-bc6e-f3e694b46deb)

**Comment by satra** [comment]
nice! also weird to see awardnumber associated with a person.

would we do it for just contributors or show a diff of any change in any section? i.e. should we treat the editable metadata as a whole. there are some parts of the metadata that are read only, and the json schema knows this. those parts are only edited by the backend service. users cannot change them.

**Comment by magland** [comment]
I was imagining that you'd be able to edit each section... but then you'd make all the changes in a single operation once ready (that's where the dandi API request would happen).

**Comment by magland** [comment]
But idk, maybe it's better to save each section individually. on the fence.

**Reply by satra** [reply]
i don't think there is a clear notion of which UX is better. just that it would be good to also identify sections that have no information.

**Comment by magland** [comment]
Another thought. I am putting this as a test page on neurosift just for convenience. But really this should be a standalone app since it does a very specific thing. It could eventually be absorbed by DANDI... but idk, pros/cons.

**Reply by satra** [reply]
even more ideally it should just be an additional ui element of the current meditor. perhaps people can switch to this "mode" versus the editing mode.

**Comment by magland** [comment]
Regarding the schema and parts that are readonly, etc. I've been going off of this spec

https://github.com/dandi/dandi-archive/blob/master/web/src/types/schema.ts

Is there a better place I should be looking for this? A json or yaml schema?

**Reply by satra** [reply]
https://github.com/dandi/schema/tree/master/releases/0.6.8

**Reply by satra** [reply]
that's the latest release.

**Reply by magland** [reply]
Okay cool. So for the readonly fields... do they just get ignored when updating metadata via api call?

**Reply by satra** [reply]
yes. the server discards/overwrites them as appropriate.

**Comment by magland** [comment]
@satra I am now using the provided dandiset.json and feeding that as context to the assistant (the entire thing is around 10k tokens, but I extract just the part needed for contributors and it's more like 2k tokens).

To test this you can prompt it with something like "Add a fake organization with a lot of sample info to highlight the schema"

![image](https://github.com/user-attachments/assets/d46dccc9-7429-46cc-a71a-2e60f3aaa5af)

**Reply by magland** [reply]
And it also does validation as part of the tool call to guarantee that it is valid.

**Comment by magland** [comment]
@satra - now you can (almost) save data back to the dandiset.

Different link:
https://neurosift.app/?p=/import

You can put in your DANDI *Staging* API key

It will show the staging dandisets owned by you. I have added you as an owner to 215759 to use as an example.

Select a dandiset.

Now you can use the chat interface to modify the contributors.

A button appears "Save changes to Dandiset Metadata"

Clicking that **almost** works to save it back to dandi staging.

The issue I believe is that CORS access needs to be added for the domain neurosift.app (localhost:3000 would also be nice). The reason I think that CORS is the only issue is that when I "Copy as curl" from the dev console of the browser and paste it into a terminal, it works.
